lib/libvfio-user.c

```diff
@@ -894,201 +894,284 @@ static int
 
-    if (msg->in.iov.iov_len < sizeof(*dirty_pages) ||
-        dirty_pages->argsz < sizeof(*dirty_pages)) {
-        vfu_log(vfu_ctx, LOG_ERR, "invalid message size %zu", msg->in.iov.iov_len);
+    if (req->argsz < msg->out.iov.iov_len) {
+        msg->out.iov.iov_len = 0;
         return ERROR_INT(EINVAL);
     }
 
-    if (vfu_ctx->migration == NULL) {
-        vfu_log(vfu_ctx, LOG_ERR, "migration not configured");
-        return ERROR_INT(ENOTSUP);
+    msg->out.iov.iov_base = calloc(1, msg->out.iov.iov_len);
+
+    if (msg->out.iov.iov_base == NULL) {
+        return ERROR_INT(ENOMEM);
     }
 
-    switch (dirty_pages->flags) {
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_START:
-        ret = dma_controller_dirty_page_logging_start(vfu_ctx->dma,
-                  migration_get_pgsize(vfu_ctx->migration));
-        break;
+    memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base, header_size);
 
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_STOP:
-        dma_controller_dirty_page_logging_stop(vfu_ctx->dma);
-        ret = 0;
-        break;
+    struct vfio_user_device_feature *res = msg->out.iov.iov_base;
 
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_GET_BITMAP:
-        ret = handle_dirty_pages_get(vfu_ctx, msg);
-        break;
+    res->argsz = msg->out.iov.iov_len;
 
-    default:
-        vfu_log(vfu_ctx, LOG_ERR, "bad flags %#x", dirty_pages->flags);
-        ret = ERROR_INT(EINVAL);
-        break;
+    dma_controller_t *dma = vfu_ctx->dma;
+
+    assert(dma != NULL);
+
+    char *bitmap = (char *)msg->out.iov.iov_base + header_size;
+
+    int ret = dma_controller_dirty_page_get(dma,
+                                            (vfu_dma_addr_t) rep->iova,
+                                            rep->length,
+                                            rep->page_size,
+                                            bitmap_size,
+                                            bitmap);
+
+    if (ret < 0) {
+        msg->out.iov.iov_len = 0;
     }
 
     return ret;
 }
 
+static int
+handle_dma_device_feature_set(vfu_ctx_t *vfu_ctx, uint32_t feature,
+                              struct vfio_user_device_feature *res)
+{
+    dma_controller_t *dma = vfu_ctx->dma;
+
+    assert(dma != NULL);
+
+    if (feature == VFIO_DEVICE_FEATURE_DMA_LOGGING_START) {
+        struct vfio_user_device_feature_dma_logging_control *ctl =
+            (void *)res->data;
+        return dma_controller_dirty_page_logging_start(dma,
+                                                       ctl->page_size);
+    } else {
+        assert(feature == VFIO_DEVICE_FEATURE_DMA_LOGGING_STOP);
+
+        dma_controller_dirty_page_logging_stop(dma);
+        return 0;
+    }
+}
+
 static int
 handle_device_feature(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
 {
     assert(vfu_ctx != NULL);
     assert(msg != NULL);
 
     if (msg->in.iov.iov_len < sizeof(struct vfio_user_device_feature)) {
-        return -EINVAL;
+        return ERROR_INT(EINVAL);
```

Comment:
spurious change

Reply:
This was fixing a legitimate bug - `handle_device_feature` is supposed to set errno otherwise error handling falls over.